Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java

```diff
@@ -2935,6 +2985,58 @@ private DataType getExpressionDefResultType(ExpressionDef expressionDef) {
         throw new IllegalArgumentException(String.format("Invalid context reference from %s context to %s context.", currentExpressionContext(), expressionDef.getContext()));
     }
 
+    /**
+     * Add an identifier to the deque to indicate that we are considering it for consideration for identifier hiding and
+     * adding a compiler warning if this is the case.
+     * <p/>
+     * For example, if an alias within an expression body has the same name as a parameter, execution would have
+     * added the parameter identifier and the next execution would consider an alias with the same name, thus resulting
+     * in a warning.
+     * <p/>
+     * Exact case matching as well as case-insensitive matching are considered.  If known, the type of the structure
+     * in question will be considered in crafting the warning message, as per the {@link Element} parameter.
+     *
+     * @param identifier The identifier belonging to the parameter, expression, function, alias, etc, to be evaluated.
+     * @param onlyOnce   Special case to deal with overloaded functions, which are out scope for hiding.
+     * @param element    The construct element, for {@link ExpressionRef}.
+     */
+    void pushIdentifierForHiding(String identifier, boolean onlyOnce, Element element) {
+        final MatchType matchType = identifiersToCheckForHiding.stream()
+                .map(innerIdentifier -> {
+                    if (innerIdentifier.equals(identifier)) {
+                        return MatchType.EXACT;
+                    }
+
+                    if (innerIdentifier.equalsIgnoreCase(identifier)) {
+                        return MatchType.CASE_IGNORED;
```

Comment:
I'm not sure this is the right implementation, a case-insensitive match isn't technically hiding. The intent of adding a warning for a case-insensitive match was actually on identifier resolution when there wasn't an exact resolution, but there is a case-insensitive resolution, so we can say something friendly like "did you mean X instead of x here?"

Reply:
My branch now outputs two distinct messages depending on whether the hiding is for an exact or case-sensitive match.

Reply:
I'm still not sure this is right, I don't think the compiler should output a warning if I just create an identifier that differs only in case, that is an intentional feature of the language and shouldn't cause a warning.

The warning should be output _only_ if a failed resolution differs from an in context identifier only by case.

In other words, this logic is running when the identifier is declared, the warning should be output on identifier resolution, not identifier declaration. Am I missing something here?

Reply:
Yes, that behavior is different than originally specified in the story as I read it. The scenario you describe would be an _error_ since it’d be a non-existent identifier. A failed resolution is a failed resolution regardless of casing as far as compilation success goes. What you’re suggesting could be implemented as a more precise error message if resolution failed (e.g. “Could not find identifier X, did you mean x?”) but it’d still be an error.  

@lukedegruchy - can we remove the case-sensitivity aspect and put that in a separate story?

Reply:
@JPercival Agreed.

Do you think this issue covers the requirement?  https://github.com/cqframework/clinical_quality_language/issues/598

Reply:
Yes, that is the story, it is incorrectly characterized there as a warning, which is the source of the confusion here, my apologies for mis-stating that in the original story definition, but yes, I think that story is the one and no need to create a separate story for this.

Reply:
I think #598 is updated to reflect that it should be a more precise error message, and then this PR does not cover that story. We’ll have a follow-on PR to cover the #598 issue separately. Does that make sense?

Reply:
I've remove the case insenstive logic and adjusted the tests accordingly.